### ISSUE TITLE
ES 5.2.x Bulk Indexing Errors

### ISSUE THREAD

**Issue by zakird** [jlinn/elasticsearch-analysis-url#20]
We're running into an issue where we can't seem to bulk index records that have fields that are analyzed using the tokenizer. The errors seem to always be of the following form:

```
[2017-02-21T18:41:16,181][DEBUG][o.e.a.s.TransportSearchAction] [es-frontend-1] [163] Failed to execute fetch phase
org.elasticsearch.transport.RemoteTransportException: [es-data-9][10.0.118.139:9300][indices:data/read/search[phase/fetch/id]]
Caused by: org.elasticsearch.search.fetch.FetchPhaseExecutionException: Fetch Failed [Failed to highlight field [parsed.extensions.authority_info_access.ocsp_urls]]
	at org.elasticsearch.search.fetch.subphase.highlight.PlainHighlighter.highlight(PlainHighlighter.java:140) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.search.fetch.subphase.highlight.HighlightPhase.hitExecute(HighlightPhase.java:124) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:163) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:489) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.action.search.SearchTransportService$13.messageReceived(SearchTransportService.java:354) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.action.search.SearchTransportService$13.messageReceived(SearchTransportService.java:351) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.transport.TcpTransport$RequestHandler.doRun(TcpTransport.java:1488) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:596) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-5.2.1.jar:5.2.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_121]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_121]
Caused by: org.elasticsearch.common.io.stream.NotSerializableExceptionWrapper: concurrent_modification_exception: null
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:901) ~[?:1.8.0_121]
	at java.util.ArrayList$Itr.next(ArrayList.java:851) ~[?:1.8.0_121]
	at org.elasticsearch.index.analysis.url.URLTokenizer.tokenize(URLTokenizer.java:176) ~[?:?]
	at org.elasticsearch.index.analysis.url.URLTokenizer.incrementToken(URLTokenizer.java:124) ~[?:?]
	at org.elasticsearch.index.analysis.url.URLTokenFilter.tokenize(URLTokenFilter.java:192) ~[?:?]
	at org.elasticsearch.index.analysis.url.URLTokenFilter.advance(URLTokenFilter.java:147) ~[?:?]
	at org.elasticsearch.index.analysis.url.URLTokenFilter.incrementToken(URLTokenFilter.java:122) ~[?:?]
	at org.apache.lucene.search.highlight.Highlighter.getBestTextFragments(Highlighter.java:216) ~[lucene-highlighter-6.4.1.jar:6.4.1 72f75b2503fa0aa4f0aff76d439874feb923bb0e - jpountz - 2017-02-01 14:44:23]
	at org.elasticsearch.search.fetch.subphase.highlight.PlainHighlighter.highlight(PlainHighlighter.java:125) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.search.fetch.subphase.highlight.HighlightPhase.hitExecute(HighlightPhase.java:124) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:163) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:489) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.action.search.SearchTransportService$13.messageReceived(SearchTransportService.java:354) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.action.search.SearchTransportService$13.messageReceived(SearchTransportService.java:351) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.transport.TcpTransport$RequestHandler.doRun(TcpTransport.java:1488) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:596) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-5.2.1.jar:5.2.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_121]
	at java.lang.Thread.run(Thread.java:745) ~[?:1.8.0_121]
```

And this will occur for random fields that use this analyzer. It never looks to occur for fields that use other tokenizers. These fields do not have any special high-lighting settings. 

This is the configuration of the filter and analyzer:
```
curl -XPUT 'localhost:9200/certificates/_settings' -d '{
  "analysis" : {
    "filter":{
      "URL":{
        "type":"url",
        "part":["protocol", "host", "port", "path", "query", "ref"],
        "url_decode":true,
        "allow_malformed":true,
        "tokenize_malformed":true
      }
    }
  }
}'
curl -XPUT 'localhost:9200/certificates/_settings' -d '{
  "analysis" : {
    "analyzer":{
      "URL":{
        "type":"custom",
        "tokenizer":"whitespace",
        "filter":["URL"]
      }
    }
  }
}'
```

I'm happy to provide any additional information that's helpful, or to help troubleshoot. Any ideas?

**Comment by zakird** [comment]
Note: This does not occur if you bulk load one record at a time. There doesn't seem to be an issue that there's multiple URL fields within the same record. It's that there are multiple records in the same bulk load request.

**Comment by zakird** [comment]
Second note: If you have multiple bulk uploader processes, it looks like this will _sometimes_ trigger, but very rarely.

**Comment by jlinn** [comment]
Thanks for reporting this; it was a tricky one.  The issue should be fixed in [v5.2.1.1](https://github.com/jlinn/elasticsearch-analysis-url/releases/tag/v5.2.1.1).

**Comment by zakird** [comment]
This was resolved a while back. Closing issues since I just happened to notice that it was still open.
